src/cswinrt/code_writers.h

```diff
@@ -7640,26 +7653,203 @@ return false;
 
     void write_delegate(writer& w, TypeDef const& type)
     {
-        if (settings.component)
-        {
-            write_authoring_metadata_type(w, type);
-            return;
-        }
-
         method_signature signature{ get_delegate_invoke(type) };
-        w.write(R"(%%%%% delegate % %(%);
+        w.write(R"(
+%%%% delegate % %(%);
 )",
-            bind<write_winrt_attribute>(type),
-            bind<write_winrt_helper_type_attribute>(type),
-            bind<write_winrt_exposed_type_attribute>(type, false),
-            bind<write_type_custom_attributes>(type, false),
+            bind<write_winrt_metadata_attribute>(type),
+            bind<write_type_custom_attributes>(type, true),
+            bind<write_comwrapper_marshaller_attribute>(type),
             internal_accessibility(),
             bind<write_projection_return_type>(signature),
             bind<write_type_name>(type, typedef_name_type::Projected, false),
             bind_list<write_projection_parameter>(", ", signature.params()));
     }
 
+    void write_delegate_impl(writer& w, TypeDef const& type)
+    {
+        auto method = get_delegate_invoke(type);
+        method_signature signature{ method };
+
+        w.write(R"(
+internal static unsafe class %Impl
+{
+    [FixedAddressValueType]
+    private static readonly %Vftbl Vftbl;
+
+    private const int S_OK = unchecked((int)0x00000000);
```

Comment:
Nit: constants should always be above all fields (applies to all other "Impl" types too)